ITE/4/README.adoc

```diff
@@ -0,0 +1,462 @@
+
+`github+https://github.com/in-toto/in-toto/pull/250`
+
+This can also instead use GitHub's API specific URI to refer to the same data.
+
+`github+https://api.github.com/repos/in-toto/in-toto/pulls/250`
+
+Both the above URIs refer to the same abstract entity - pull request 250 to
+in-toto's reference implementation. However, taken directly, they correspond to
+different data. Therefore, the implementing system must resolve the URI and thus
+the artifact in a consistent manner. We discuss this and other sources of
+non-determinism in the Appendix.
+
+An example of the product field for link metadata with a generic URI is as
+follows:
+
+```
+"products": {
+    "github+https://api.github.com/repos/in-toto/in-toto/pulls/250": {
+        "sha256": "7ee0617..."
+    },
+    "...": ...
+}
+```
+
+Similarly, a generic URI can be used to refer to other abstract entities like an
+SPDX element. The URI can be used to directly fetch the corresponding documents
+from a specified location. For in-toto metadata that are embedded within SPDX
+documents, the URI can also be used to refer to other entities elsewhere in the
+document. Resolving such URIs that point to local entities elsewhere in the SPDX
+document is an implementation level detail - but it's clear that the URI must
+contain all the information necessary to identify the entity in that context.
+
+It is important to note that this ITE doesn't mandate any specific generic URI
+tokens that all conforming implementations MUST support. Instead, it merely
+specifies what any generic URI must look like, and what rules the system MUST
+follow.
+
+in-toto artifact rules are a means to specify what artifacts the project owner
+expects in every step of the supply chain in the in-toto layout. in-toto
+provides the following rules that together attempt to describe the different
+types of operations or actions that supply chain functionaries carry out.
+
+*MATCH Rule*
+
+`MATCH <pattern> [IN <source-path-prefix>] WITH (MATERIALS|PRODUCTS) [IN
+<destination-path-prefix>] FROM <step>`
+
+The `MATCH` rule is a convenient way to match artifacts (either in materials or
+products depending on where the rule is specified) with artifacts from other
+steps in the supply chain, allowing owners to establish a flow of artifacts
+between steps in the software supply chain.
+
+The current verification workflow for the `MATCH` rule compares the
+cryptographic hashes of the relevant artifacts from the respective link files.
+This is a straightforward comparison and does not entail resolving the URIs in
+any form. The `IN` clauses which are used to specify path prefixes are
+subtracted from the URI strings. Therefore, we conclude that the functioning of
+the rule doesn't change with this ITE. An example `MATCH` rule with generic URIs
+is as follows:
+
+`MATCH commit/* IN github+https://github.com/in-toto/in-toto/ WITH PRODUCTS IN
+github+https://github.com/in-toto/in-toto FROM merge-pull-request-250`
```

Comment:
How do you deal with the fact that the content at a URI may change over time?  Is this intended to be allowed?  Are URIs meant to point to immutable content?

Reply:
No, URIs don't point only at immutable content. I think this overlaps with the point above about hashing changing content. The in-toto specification doesn't keep up with changes to artifacts with time. I think the general expectation is that metadata is versioned / re-generated or perhaps the changes are themselves captured with link attestations, but I'd like confirmation. Our analysis was that the temporal aspects of in-toto metadata are not in scope of this ITE.

Reply:
> No, URIs don't point only at immutable content. I think this overlaps with the point above about hashing changing content. 

Yes, it does.  Okay, this seems like an important thing to be very clear about early in the document.  What happens if, for example, github changes the format of the data they present at a URI?  Will client verification / signing code need to change?  

> The in-toto specification doesn't keep up with changes to artifacts with time.  I think the general expectation is that metadata is versioned / re-generated or perhaps the changes are themselves captured with link attestations, but I'd like confirmation. 

Can you give an example of what you mean so I can be sure this is clear?

> Our analysis was that the temporal aspects of in-toto metadata are not in scope of this ITE.

These issues become much more important here because the data at URIs is likely to change over time.  I do think you're right that an ITE that covers temporal aspects would help this ITE tremendously.

Reply:
> Can you give an example of what you mean so I can be sure this is clear?

I mean that if the contents of a file in the filesystem that is recorded as an artifact in in-toto links change, then verification of existing metadata will fail. Either the change itself is recorded as a step and we use its link to validate the change, or the functionary must replace existing metadata with the updated ones, capturing the new hash(es) of the artifact.

> What happens if, for example, github changes the format of the data they present at a URI? Will client verification / signing code need to change?

ITE-4 currently specifies that the data represented as an abstract entity must be resolved and then hashed. So a change in format of the data _could_ require a change in how the URI is resolved to provide hashable data. But this resolving operation should be separate from the verification and signing code. Does that seem sane? It may also just be a difference in semantics, so I suppose the answer is yes, a change of some nature may be necessary.

Reply:
A change in either content or format seems to generalize to the same problem. I think you describe this well in the 3rd paragraph of the Security section below.

Maybe you can somehow add your suggestion that  *"the change itself is recorded as a step and we use its link to validate the change, or the functionary must replace existing metadata with the updated ones"*.